---
title: "Pf3k PNG Gene of Interest"
author: "Somya Menhra and Myo Naung" 
date: "09/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(ggplot2)
library(knitr)
library(kableExtra)
library(SeqArray)
library(dplyr)
library(tidyr)
library(reshape)
library(ape)
showfile.gds(closeall = TRUE)
```

## Input Data


```{r data}

# Data input 


pf3kv5 <- seqOpen("filepath/WGS_GATK_Processed_and_SNPeff_annotated_file.gds", readonly = TRUE)

# geographic metadata
metadata <- read.delim("filepath/metadata_with_sample_ID.txt",
                       header=TRUE, stringsAsFactors=FALSE, na.strings=c(""))
rownames(metadata) <- metadata$sample

# target antigens
antigens <- read.delim("filepath/antigen_list.txt", header=TRUE, stringsAsFactors=FALSE)
rownames(antigens) <- antigens$Gene.ID

# functional annotations for varaints passing GATK VQSR + further hard-filtering
# based on GATK parameters and missingness
func_annotations <- readr::read_rds("filepath/pf3kv5_functional_annotations_filtered.rds")
gatk_annotations <- readr::read_rds("filepath/pf3kv5_gatk_metrics.rds")

# hard_filtered <- func_annotations %>% select(chr, pos) %>% distinct %>% arrange(chr, pos)
# write.table(hard_filtered, "pf3k5_hard_filtered_loci.txt", col.names = FALSE, row.names=FALSE, sep="\t", quote=FALSE)

# MOI for samples passing missingness and depth of coverage filters
# MOI was run using https://github.com/bahlolab/moimix.
# MOI was determined based on Fws scores
# MOI 1 = fws>0.90
# MOI 2 = fws<=0.90 & moi$fws>0.80

moi <- readr::read_rds("path/moimix_output/pf3kv5_fws.rds")
moi1_samples <- moi$sample.id[moi$fws>0.90]
moi2_samples <- moi$sample.id[moi$fws<=0.90 & moi$fws>0.80] 
keep_samples <- moi$sample.id[!is.na(moi$moi)]

BIN_SIZE <- 30 # for SNP density plots
MOI1_MIN_DEPTH <- 5 # minimum depth of coverage per site for sample with MOI 1
MOI2_MIN_DEPTH <- 10 # minimum depth of coverage per site for sample with MOI 2
HET_THRESH <- 0.2   # minimum minor allele proportion for a heterozygous call to be made
MINOR_PROP_THRESH <- 0.4 # we differentiate the major and minor haplotypes by
                         # read depth; the mean minor allele frequency must be below
                         # this threshold for the major and minor haplotypes to be
                         # differentiated properly

```

## Haplotype Extraction
```{r hap}
# EXTRACT MAJOR AND MINOR HAPLOTYPES FOR SAMPLES WITH MOI==2
# returns a list of dataframes, one for each sample, encoding haplotypes
# over a given set of loci; maj_clone and min_clone give the allele calls
# for each clone, while maf and baf denote the major and b-allele freqs
# het_thresh is the minimum b allele frequency for a true polyclonal call
# loci with read depth < min_depth are filtered out
moi2_haplotypes <- function(var_subset, sample_subset, mgen, 
                            het_thresh, min_depth) {
  
  seqSetFilter(mgen, variant.id=var_subset, sample.id=sample_subset)
  
  haplotypes <- list()
  sample_hap <- data.frame(variant.id=var_subset, maj_clone=NA, min_clone=NA,
                           maj_clone_bin=NA, min_clone_bin=NA, maf=NA, baf=NA, het=FALSE)
  
  # initialise list of haplotypes, one df for each sample
  for (sample in sample_subset) {
    haplotypes[[sample]] <- sample_hap
  }
  
  # pick out allele depths + allele calls
  all_depths <- seqGetData(mgen, "annotation/format/AD")
  all_calls <- seqGetData(mgen, "allele") %>% strsplit(",")
  num_per_locus <- all_depths$length
  depth_per_locus <- all_depths$data
  
  variant_num <- 1
  curr <- 0
  
  # iterate over each variant
  while (variant_num <= length(num_per_locus)) {
    num <- num_per_locus[variant_num]

    # extract major/minor clones and maf/baf for each sample
    for (i in 1:length(sample_subset)) {
      depths <- depth_per_locus[i,(curr+1):(curr+num)]
      calls <- order(-depths)
      if (sum(depths)>min_depth) {
        haplotypes[[i]][variant_num, "maj_clone"] <- all_calls[[variant_num]][calls[1]]
        haplotypes[[i]][variant_num, "min_clone"] <- all_calls[[variant_num]][calls[2]]
        haplotypes[[i]][variant_num, "maj_clone_bin"] <- calls[1]
        haplotypes[[i]][variant_num, "min_clone_bin"] <- calls[2]
        haplotypes[[i]][variant_num, "maf"] <- depths[calls[1]]/sum(depths)
        haplotypes[[i]][variant_num, "baf"] <- depths[calls[2]]/sum(depths)
        if (haplotypes[[i]][variant_num, "baf"]<het_thresh) {
          haplotypes[[i]][variant_num, "min_clone"] <- haplotypes[[i]][variant_num, "maj_clone"]
          haplotypes[[i]][variant_num, "min_clone_bin"] <- haplotypes[[i]][variant_num, "maj_clone_bin"]
        } else {
          haplotypes[[i]][variant_num, "het"] <- TRUE
        }
      } else {
        haplotypes[[i]][variant_num, "min_clone_bin"] <- haplotypes[[i]][variant_num, "maj_clone_bin"] <- 0
      }
    }
    
    variant_num <- variant_num + 1
    curr <- curr+num  
  }
  
  return(haplotypes)
}

# EXTRACT SEQUENCES AND CLONAL PROPORTIONS FOR SAMPLES WITH MOI==2
# returns a list of dataframes; clones contains the major and minor
# clone proportions for each sample, while seqs contains allele sequences
# for the major and (if applicable) minor clones in each sample
# we determine the major and minor haplotypes using read depths; the
# minor clone proportion must be below min_prop_thresh for haplotypes
# to be pieced together appropriately
moi2_sequences <- function(variants, samples, haplotypes, min_prop_thresh) {
  n_samp <- length(samples)
  clones <- data.frame(row.names=samples, maj_prop=rep(1, n_samp), 
                       min_prop=rep(0, n_samp), retained=FALSE)
  seqs <- data.frame(row.names=variants)
  
  # for true het calls, we calculate mean(maf) = maj_prop and mean(baf) = minor_prop
  # and output the allele sequences for both the major and minor clones
  # if a sample has no true het calls, we output the major clone only
  for (sample in samples) {
    if (any(haplotypes[[sample]]$het)) {
      het_calls <- haplotypes[[sample]] %>% subset(het==TRUE)
      clones[sample, "maj_prop"] <- mean(het_calls$maf)
      clones[sample, "min_prop"] <- mean(het_calls$baf)
      if (clones[sample, "min_prop"]<min_prop_thresh){
        min_name <- paste0(sample, "_min")
        seqs <- cbind(haplotypes[[sample]]$maj_clone,
                      haplotypes[[sample]]$min_clone, seqs)
        colnames(seqs)[1] <- paste0(sample, "_maj")
        colnames(seqs)[2] <- paste0(sample, "_min")
        clones[sample, "retained"] <- TRUE
      }
    } else {
      seqs <- cbind(haplotypes[[sample]]$maj_clone, seqs)
      colnames(seqs)[1] <- paste0(sample, "_maj")
    }
  }
  return(list(clones=clones, seqs=seqs))
}

#==========================================================================

# EXTRACT SEQUENCES FOR SAMPLES WITH MOI==1
# returns a dataframe of allele calls for a given set of variants and
# samples, given a per-site constraint on the depth of coverage
moi1_sequences <- function(variants, samples, mgen, min_depth) {
  n_sample = length(samples)
  n_var = length(variants)
  
  seqs <- gen_bin <- setNames(data.frame(matrix(ncol=n_sample, nrow=n_var), 
                              row.names=variants), samples)

  seqSetFilter(mgen, variant.id=variants, sample.id=samples)

  # extract genotypes (0=ref, 1,2... are alt alleles)
  gen <- seqGetData(mgen, "genotype") %>% matrix(ncol=n_var, nrow=n_sample)
  
  # pick out allele calls
  all_calls <- seqGetData(mgen, "allele") %>% strsplit(",")
  
  # pick out read depth at each site
  cov_depth <- seqGetData(mgen, "annotation/format/DP")$data
  cov_depth[is.na(cov_depth)] <- 0
  
  # determine the allele called for each sample, given the depth
  # of coverage at any given site is at least min_depth
  for (var_num in 1:n_var) {
    for (samp_num in 1:n_sample) {
      if (cov_depth[samp_num, var_num]>=min_depth) {
         seqs[var_num, samp_num] <- all_calls[[var_num]][gen[samp_num,var_num]+1]
         gen_bin[var_num, samp_num] <- gen[samp_num,var_num]+1
      } else {
         gen_bin[var_num, samp_num] <- 0
      }
    }
  }

  return(list(seqs=seqs, genotypes=gen_bin))
}

#==========================================================================
# FORMAT CONVERSION

# generate a ped file ped_out (with family ID given by country of origin, and
# maternal ID, paternal ID and phenotype all set to 0) and a map_file for
# samples with MOI==1 or MOI==2 across a subset of variants
generate_pedmap <- function(variants, moi1_haps, moi2_haps, metadata, ped_out, map_out) {
  cat("", file=ped_out, append=FALSE)
  for (sample in names(moi2_haps)) {
    loc <- moi2_haps[[sample]] %>% transmute(loc=paste0("\t", maj_clone_bin, "\t", min_clone_bin))
    cat(paste0(metadata[sample,]$country, "\t", sample, "\t", "0\t0\t2\t0", 
               paste(loc$loc, collapse=""), "\n"), file=ped_out, append=TRUE)
  }
  for (sample in colnames(moi1_haps[["genotypes"]])) {
    cat(paste0(metadata[sample,]$country, "\t", sample, "\t", "0\t0\t1\t0\t", 
               paste(rep(moi1_haps[["genotypes"]][, sample], each=2), collapse="\t"), 
               "\n"), file=ped_out, append=TRUE)
  }
  variants %>% mutate(cM=pos/15000) %>% select(chr, variant_id, cM, pos) %>% 
    write.table(file=map_out, row.names = FALSE, col.names = FALSE, sep="\t", quote=FALSE)
}

# we take the complementary base to the allele called in the VCE when annotating the CDS fasta file
comp_base <- function(base) {
  if (toupper(base)=="A") return("T")
  else if (toupper(base)=="T") return("A")
  else if (toupper(base)=="C") return("G")
  else if (toupper(base)=="G") return("C")
  else return(NA)
}

# generate a fasta file fasta_out, containing both major and minor haplotypes
generate_fasta <- function(cds_seq, cds_pos, moi1_haps, moi2_seqs, fasta_out, direction) {
  cat("", file=fasta_out, append=FALSE)
  seqs <- cbind(moi1_haps[["seqs"]], moi2_seqs[["seqs"]]) %>% apply(1:2, as.character)
  seqs[is.na(seqs)] <- "N"
  for (sample in colnames(seqs)) {
    cat(paste0(">", sample, "\n"), file=fasta_out, append=TRUE)
    sample_seq <- cds_seq
    for (i in 1:length(cds_pos)) sample_seq[cds_pos[i]] <- ifelse(as.character(direction)=="NEGATIVE", comp_base(seqs[i, sample]), seqs[i, sample])
    cat(paste0(paste(sample_seq, collapse=""), "\n"), file=fasta_out, append=TRUE)
  }
}

```

```{r by_gene, fig.height=4, fig.width=12, results='asis', message=FALSE}
clonal_props <- data.frame(samples=as.character(moi2_samples))
for (gene in antigens$Gene.ID) {
  cat("\n******  \n")
  cat(paste0("\n###", gene, "  \n"))
  
  # extract coding SNPs in gene of interest
  snps <- func_annotations %>% 
    subset(gene_id==gene & 
             (annotation=="missense_variant" | annotation=="synonymous_variant" |
              annotation=="missense_variant&splice_region_variant" |
              annotation=="splice_region_variant&synonymous_variant" |
              annotation=="stop_gained" | annotation=="stop_lost" |
              annotation=="stop_lost&splice_region_variant" |
              annotation=="start_lost" | annotation=="stop_gained&splice_region_variant" |
              annotation=="splice_region_variant&stop_retained_variant" |
              annotation=="stop_retained_variant"))
  if (nrow(snps) > 0) {  
  # visualise the distribution of SNPs across the gene
  gene_var <- snps %>% select(chr, variant_id, pos, alt, CDS) %>% distinct %>%
    group_by(chr, variant_id, pos, CDS) %>% summarise(alleles=(n()+1)) %>% 
    extract(CDS, into=c("pos_CDS", "len_CDS"), regex="([0-9]+)/([0-9]+)", convert=TRUE) %>%
    as.data.frame
  snp_bins <- snps %>% mutate(bins=ceiling((pos-antigens[gene, "start"])/BIN_SIZE)) %>%
    group_by(bins) %>% summarise(count=n())
  snp_pos <- ggplot() + 
    geom_segment(data=gene_var, aes(x=pos, xend=pos, y=0, yend=1, 
                                    colour=as.character(alleles)), alpha=0.8) +
    scale_colour_manual(values=c("red", "green", "blue"), 
                        breaks=c("2", "3", "4"), name="Number of alleles") +
    xlab("Position (bp)") + xlim(c(antigens[gene, "start"], antigens[gene, "end"])) +
    ggtitle(paste0("SNP Loci for ", gene)) + theme_classic() + 
    theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(),
          axis.title.y = element_blank(), axis.line.y = element_blank(),
          plot.title=element_text(hjust=0.5, face="bold"))

  # summary of coding SNPs across gene
  cat(paste0("Number of SNPs: ", nrow(gene_var)))
  cat("\n")
  print(kable(snps, caption = paste0("Coding SNPs for ", gene)) %>% 
          kable_styling() %>% scroll_box(width="100%", height="300px"))
  cat("\n  ")
  print(snp_pos)
  
  # take a subset of data -> samples with MOI 1 or 2 + coding variants in gene 
  # of interest only
  seqSetFilter(pf3kv5, variant.id = gene_var$variant_id, sample.id = as.character(keep_samples))
  cat("  \n")

  # write data to a vcf file
  seqGDS2VCF(pf3kv5, paste0("filepath/vcf/", gene, "_coding.vcf"),
             fmt.var=c("AD", "DP", "GQ", "PL", "RGQ", "SB"))
  
  # extract haplotypes for samples with MOI 1
  moi1_haps <- moi1_sequences(gene_var$variant_id, as.character(moi1_samples), pf3kv5, MOI1_MIN_DEPTH)
  
  # extract haplotypes for samples with MOI 2
  moi2_haps <- moi2_haplotypes(gene_var$variant_id, as.character(moi2_samples), pf3kv5, HET_THRESH, MOI2_MIN_DEPTH)
  moi2_seqs <- moi2_sequences(gene_var$variant_id, as.character(moi2_samples), moi2_haps, MINOR_PROP_THRESH)
  
  # extract reference coding sequence for gene from fasta file
  cds_seq <- read.dna(paste0("filepath/cds_fasta", gene, ".1.fasta"), format="fasta") %>% as.character %>% as.vector %>% toupper
  # generate a fasta file containing coding sequences for each clone 
  generate_fasta(cds_seq, gene_var$pos_CDS, moi1_haps, moi2_seqs, paste0("filepath/fasta_file_output/", gene, "_CDS.fasta"), antigens[gene, ]$dir)


  # keep track of clonal proportions for samples with MOI 2
  clonal_props <- cbind(moi2_seqs[["clones"]]$min_prop, clonal_props)
  colnames(clonal_props)[1] <- gene
  
  # graph a boxplot of minor clone proportions for the gene
  boxplot(moi2_seqs[["clones"]]$min_prop, horizontal=TRUE, ylim=c(0, 0.5), xlab="Minor Clone Proportion (mean BAF across het loci)", main=paste0("Minor Clone Proportions for ", gene)) 
  
  # generate a pedmap file if there are more than 10 variants across the gene
  if (nrow(gene_var)>10) {
      generate_pedmap(gene_var, moi1_haps, moi2_haps, metadata, paste0("filepath/pedmap/", gene, "_coding.ped"), paste0("filepath/pedmap/", gene, "_coding.map"))
  }
  
  seqResetFilter(pf3kv5)
  }
}

```

```{r by_sample, results='asis', fig.height=4, fig.width=12}
cat("\n******  \n")
cat(paste0("\n###", "Minor Clone Proportions by Sample\n"))
kable(clonal_props) %>% kable_styling() %>% scroll_box(width="100%", height="300px")
clonal_props <- clonal_props %>% tibble::column_to_rownames(var="samples")

for (sample in rownames(clonal_props)) {
  boxplot(as.numeric(clonal_props[sample,]), horizontal=TRUE, ylim=c(0, 0.5), xlab="Minor Clone Proportion (mean BAF across het loci)", main=paste0("Minor Clone Proportions for ", sample)) 
}
```
